csharp/src/iceboxnet/ServiceManager.cs

```diff
@@ -2,195 +2,142 @@
-                _pendingStatusChanges = true;
+                throw new NoSuchServiceException();
             }
 
-            bool started = false;
             try
             {
-                Debug.Assert(info.Service != null);
-                Communicator? communicator = info.Communicator ?? _sharedCommunicator;
-                Debug.Assert(communicator != null);
-                info.Service.Start(info.Name, communicator, info.Args);
-                started = true;
+                await info.StartServiceAsync(_sharedCommunicator, cancel);
+
+                // We call ServiceStarted even if this call does not start the service - this is a way to refresh
+                // the status of this service in the observers.
+                ServiceStarted(name);
             }
             catch (Exception ex)
             {
-                _logger.Warning($"IceBox.ServiceManager: exception while starting service `{info.Name}':\n{ex}");
-            }
-
-            lock (_mutex)
-            {
-                if (started)
-                {
-                    info.Status = ServiceStatus.Started;
-                    ServiceStarted(name, _observers);
-                }
-                else
-                {
-                    info.Status = ServiceStatus.Stopped;
-                }
-                _pendingStatusChanges = false;
-                Monitor.PulseAll(_mutex);
+                _logger.Warning($"IceBox.ServiceManager: exception while starting service `{name}':\n{ex}");
+                throw new ServiceException($"service {name} failed to start", ex);
             }
         }
 
-        public void StopService(string name, Current current, CancellationToken cancel)
+        public async ValueTask StopServiceAsync(string name, Current current, CancellationToken cancel)
         {
-            ServiceInfo? info;
-            lock (_mutex)
+            ServiceInfo? info = _services.Find(service => service.Name == name);
+            if (info == null)
             {
-                // Search would be more efficient if services were contained in a map, but order is required for
-                // shutdown.
-                info = _services.Find(service => service.Name.Equals(name));
-                if (info == null)
-                {
-                    throw new NoSuchServiceException();
-                }
-
-                if (info.Status != ServiceStatus.Started)
-                {
-                    throw new AlreadyStoppedException();
-                }
-                _pendingStatusChanges = true;
+                throw new NoSuchServiceException();
             }
 
-            bool stopped = false;
             try
             {
-                Debug.Assert(info.Service != null);
-                info.Service.Stop();
-                stopped = true;
+                await info.StopServiceAsync();
+
+                // We notify the observers even if this call did not actually stop the service.
```

Comment:
Perhaps add `We notify the observers even if this call did not actually stop the service because it was already stopped`.

Reply:
Fixed as suggested.